parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema16To17.hs

```diff
@@ -0,0 +1,225 @@
+      -- trying to display some other error.
+      [[Sqlite.SQLData]]
+  | MissingRootBranch
+  deriving stock (Show)
+  deriving anyclass (Exception)
+
+addCausalHashesToProjectBranches :: Sqlite.Transaction ()
+addCausalHashesToProjectBranches = do
+  Debug.debugLogM Debug.Migration "Creating new_project_branch"
+  -- Create the new version of the project_branch table with the causal_hash_id column.
+  Sqlite.execute
+    [Sqlite.sql|
+CREATE TABLE new_project_branch (
+  project_id uuid NOT NULL REFERENCES project (id),
+  branch_id uuid NOT NULL,
+  name text NOT NULL,
+  causal_hash_id integer NOT NULL REFERENCES causal(self_hash_id),
+
+  primary key (project_id, branch_id),
+
+  unique (project_id, name)
+)
+without rowid;
+|]
+  rootCausalHashId <- expectNamespaceRoot
+  rootCh <- Q.expectCausalHash rootCausalHashId
+  projectsRoot <- Codebase.getShallowCausalAtPathFromRootHash rootCh (Path.singleton $ projectsNameSegment) >>= V2Causal.value
+  ifor_ (V2Branch.children projectsRoot) \projectIdNS projectsCausal -> do
+    projectId <- case projectIdNS of
+      UUIDNameSegment projectIdUUID -> pure $ ProjectId projectIdUUID
+      _ -> error $ "Invalid Project Id NameSegment:" <> show projectIdNS
+    Debug.debugM Debug.Migration "Migrating project" projectId
+    projectsBranch <- V2Causal.value projectsCausal
+    case (Map.lookup branchesNameSegment $ V2Branch.children projectsBranch) of
+      Nothing -> pure ()
+      Just branchesCausal -> do
+        branchesBranch <- V2Causal.value branchesCausal
+        ifor_ (V2Branch.children branchesBranch) \branchIdNS projectBranchCausal -> void . runMaybeT $ do
+          projectBranchId <- case branchIdNS of
+            UUIDNameSegment branchIdUUID -> pure $ ProjectBranchId branchIdUUID
+            _ -> error $ "Invalid Branch Id NameSegment:" <> show branchIdNS
+          Debug.debugM Debug.Migration "Migrating project branch" projectBranchId
+          let branchCausalHash = V2Causal.causalHash projectBranchCausal
+          causalHashId <- lift $ Q.expectCausalHashIdByCausalHash branchCausalHash
+          branchName <-
+            MaybeT $
+              Sqlite.queryMaybeCol @ProjectBranchName
+                [Sqlite.sql|
+                  SELECT project_branch.name
+                  FROM project_branch
+                  WHERE
+                    project_branch.project_id = :projectId
+                    AND project_branch.branch_id = :projectBranchId
+                |]
+          -- Insert the full project branch with HEAD into the new table
+          lift $
+            Sqlite.execute
+              [Sqlite.sql|
+              INSERT INTO new_project_branch (project_id, branch_id, name, causal_hash_id)
+                VALUES (:projectId, :projectBranchId, :branchName, :causalHashId)
+            |]
+
+  Debug.debugLogM Debug.Migration "Deleting orphaned project branch data"
+  -- Delete any project branch data that don't have a matching branch in the current root.
```

Comment:
Delete any projects/branches which don't have a corresponding namespace in the root, generally there shouldn't be any, but if they don't have any code for a branch it's useless anyways.